modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryWithConcurrentRollbackTest.java

```diff
@@ -258,6 +264,129 @@ else if (g1Keys.contains(key))
+     * Use several attempts to reproduce the race condition.
+     * <p>
+     * Expected result: transaction is finished on both backup nodes and the partition map exchange is completed as well.
+     */
+    @Test
+    public void testRecoveryNotDeadLockOnPrimaryFail() throws Exception {
+        backups = 2;
+        persistence = false;
+
+        for (int iter = 0; iter < 100; iter++) {
+            stopAllGrids();
+
+            log.info("iteration=" + iter);
+
+            final IgniteEx grid0 = startGrid(0);
+
+            final IgniteEx grid1 = startGrid(1, (UnaryOperator<IgniteConfiguration>)cfg ->
+                    cfg.setSystemThreadPoolSize(1).setStripedPoolSize(1));
+
+            final IgniteEx grid2 = startGrid(2);
+
+            grid0.cluster().state(ACTIVE);
+
+            final IgniteCache<Object, Object> cache = grid2.cache(DEFAULT_CACHE_NAME);
+
+            final Integer g2Key = primaryKeys(cache, 1, 0).get(0);
+
+            List<IgniteInternalTx> txs0 = null;
+            List<IgniteInternalTx> txs1 = null;
+            List<IgniteInternalTx> txs2;
+
+            final CountDownLatch grid1BlockLatch = new CountDownLatch(1);
+
+            final CountDownLatch grid1NodeLeftEventLatch = new CountDownLatch(1);
+
+            int[] stripeHolder = new int[1];
+
+            try (final Transaction tx = grid2.transactions().txStart(PESSIMISTIC, REPEATABLE_READ)) {
+                cache.put(g2Key, Boolean.TRUE);
+
+                TransactionProxyImpl<?, ?> p = (TransactionProxyImpl<?, ?>)tx;
+
+                p.tx().prepare(true);
+
+                txs0 = txs(grid0);
+                txs1 = txs(grid1);
+                txs2 = txs(grid2);
+
+                assertTrue(txs0.size() == 1);
+                assertTrue(txs1.size() == 1);
+                assertTrue(txs2.size() == 1);
+
+                grid1.events().localListen(new PE() {
+                    @Override public boolean apply(Event evt) {
+                        grid1NodeLeftEventLatch.countDown();
+
+                        return true;
+                    }
+                }, EventType.EVT_NODE_LEFT);
+
+                // Block recovery procedure processing on grid1
+                grid1.context().pools().getSystemExecutorService().execute(() -> U.awaitQuiet(grid1BlockLatch));
+
+                int stripe = U.safeAbs(p.tx().xidVersion().hashCode());
+
+                stripeHolder[0] = stripe;
+
+                // Block stripe tx recovery request processing on grid1
+                grid1.context().pools().getStripedExecutorService().execute(stripe, () -> U.awaitQuiet(grid1BlockLatch));
+
+                // Prevent finish request processing on grid0
+                spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid0.name());
+
+                // Prevent finish request processing on grid1
+                spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid1.name());
+
+                runAsync(() -> {
+                    grid2.close();
```

Comment:
newline missed betwen the semantic units

Reply:
FIxed